File: packages/sp_shared/lib/src/models/episode_data.dart
/// Interface for episode data used by SmartPlaylist resolvers and extractors.
///
/// Abstracts away the storage layer (Drift, RSS parsed data, etc.)
/// so resolvers work with any episode source.
abstract interface class EpisodeData {
  /// Unique episode identifier.
  int get id;

  /// Episode title.
  String get title;

  /// Episode description (optional).
  String? get description;

  /// Season number from RSS metadata (optional).
  int? get seasonNumber;

  /// Episode number from RSS metadata (optional).
  int? get episodeNumber;

  /// Publication date (optional).
  DateTime? get publishedAt;

  /// Episode artwork URL (optional).
  String? get imageUrl;
}

/// Simple implementation of [EpisodeData] for testing and web service use.
final class SimpleEpisodeData implements EpisodeData {
  const SimpleEpisodeData({
    required this.id,
    required this.title,
    this.description,
    this.seasonNumber,
    this.episodeNumber,
    this.publishedAt,
    this.imageUrl,
  });

  @override
  final int id;

  @override
  final String title;

  @override
  final String? description;

  @override
  final int? seasonNumber;

  @override
  final int? episodeNumber;

  @override
  final DateTime? publishedAt;

  @override
  final String? imageUrl;
}

File: crates/sp_core/src/models/episode_data.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Interface for episode data used by resolvers and extractors.
pub trait EpisodeData {
    fn id(&self) -> i64;
    fn title(&self) -> &str;
    fn description(&self) -> Option<&str>;
    fn season_number(&self) -> Option<i32>;
    fn episode_number(&self) -> Option<i32>;
    fn published_at(&self) -> Option<DateTime<Utc>>;
    fn image_url(&self) -> Option<&str>;
}

/// Simple implementation of [EpisodeData] for testing and web service use.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SimpleEpisodeData {
    pub id: i64,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub season_number: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub episode_number: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub published_at: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image_url: Option<String>,
}

impl EpisodeData for SimpleEpisodeData {
    fn id(&self) -> i64 {
        self.id
    }

    fn title(&self) -> &str {
        &self.title
    }

    fn description(&self) -> Option<&str> {
        self.description.as_deref()
    }

    fn season_number(&self) -> Option<i32> {
        self.season_number
    }

    fn episode_number(&self) -> Option<i32> {
        self.episode_number
    }

    fn published_at(&self) -> Option<DateTime<Utc>> {
        self.published_at
    }

    fn image_url(&self) -> Option<&str> {
        self.image_url.as_deref()
    }
}
